### PR TITLE
MM-54580: Fix panic in jobs/base_workers.go

### PR DESCRIPTION
**Pull request by agnivade**
If GetJob fails, we would return a nil job and then try to log
with that. To prevent this, we keep a reference to the old job
and log with that in the error case.

https://mattermost.atlassian.net/browse/MM-54580

```release-note
Fix a panic where a simple worker would crash if if failed to get a job.
```

### CODE REVIEW

File: server/channels/jobs/base_workers.go [hunk 1]
```diff
@@ -83,11 +83,14 @@ func (worker *SimpleWorker) DoJob(job *model.Job) {
 	c := request.EmptyContext(worker.logger)
 
 	var appErr *model.AppError
+	// We keep the reference to the old job because GetJob might return nil.
+	oldJob := job
 	// We get the job again because ClaimJob changes the job status.
 	job, appErr = worker.jobServer.GetJob(c, job.Id)
 	if appErr != nil {
```
Comment (hanzei) [inline server/channels/jobs/base_workers.go:90]: 1/5 that it's simpler to write:
```golang
// We get the job again because ClaimJob changes the job status.
newJob, appErr := worker.jobServer.GetJob(c, job.Id)
if appErr != nil {
		job.Logger.Error("SimpleWorker: job execution error", mlog.Err(appErr))
		worker.setJobError(job, appErr)
}
job = newJob
```

This way, the error handling pattern stays the same.

File: server/channels/jobs/base_workers_test.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	"testing"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSimpleWorkerPanic(t *testing.T) {
+	jobServer, mockStore, mockMetrics := makeJobServer(t)
```
Comment (hanzei) [inline server/channels/jobs/base_workers_test.go:17]: Should `makeJobServer` also assign a logger to the returned `JobServer`?

File: server/channels/jobs/base_workers_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	mockStore.JobStore.On("UpdateStatusOptimistically", "job_id", model.JobStatusPending, model.JobStatusInProgress).Return(true, nil)
+	mockStore.JobStore.On("UpdateOptimistically", mock.AnythingOfType("*model.Job"), model.JobStatusInProgress).Return(true, nil)
+	mockStore.JobStore.On("Get", mock.AnythingOfType("*request.Context"), "job_id").Return(nil, errors.New("test"))
+	mockMetrics.On("IncrementJobActive", "job_type")
+	mockMetrics.On("DecrementJobActive", "job_type")
+	sWorker := NewSimpleWorker("test", jobServer, exec, isEnabled)
+
+	require.NotPanics(t, func() {
+		sWorker.DoJob(job)
+	})
```
Comment (hanzei) [inline server/channels/jobs/base_workers_test.go:44]: Great job also adding a test :tada:

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks, Agniva, for fixing my mistakes :+1:

https://github.com/mattermost/mattermost/blob/cd9e0d7509e3484b1b006f78537253bfe0d476fa/server/channels/jobs/s3_path_migration/s3_path_migration.go#L128-L133 is another piece that can panic. Should we also fix it as part of this PR?

**Comment by agnivade** [comment]
> Thanks, Agniva, for fixing my mistakes 👍

This was there before your refactor as well :stuck_out_tongue: Not your fault. :smile: 

> is another piece that can panic. Should we also fix it as part of this PR?

Ah good one. This migration is actually completed in cloud, so there's not much benefit fixing it now.

File: server/channels/jobs/base_workers.go (hunk 1 above)
Comment (agnivade) [reply server/channels/jobs/base_workers.go:90]: Ah yes, much better.

File: server/channels/jobs/base_workers_test.go (hunk 2 above)
Comment (agnivade) [reply server/channels/jobs/base_workers_test.go:17]: I tried that before, but it was throwing some error. Let me try again.

**Review by hanzei — APPROVED** [review]
Nice :+1: 

And yes, you are right, the panic wasn't actually introduce by me :sweat_smile:

**Review by BenCookie95 — APPROVED** [review]

**Comment by agnivade** [comment]
@amyblais - This would need a manual CP to 8.1 ESR. I'll send a PR.
